libtenzir/builtins/contexts/geoip.cpp

```diff
@@ -0,0 +1,485 @@
+      }
+      if (result.found_entry) {
+        status = MMDB_get_entry_data_list(&result.entry, &entry_data_list);
+        if (status != MMDB_SUCCESS) {
+          return caf::make_error(
+            ec::lookup_error, fmt::format("error looking up IP address '{}' in "
+                                          "GeoIP database: {}",
+                                          ip_string, MMDB_strerror(status)));
+        }
+        auto* entry_data_list_it = entry_data_list;
+        auto output = record{};
+        entry_data_list_it
+          = entry_data_list_to_record(entry_data_list_it, &status, output);
+        if (status != MMDB_SUCCESS) {
+          return caf::make_error(
+            ec::lookup_error, fmt::format("error looking up IP address '{}' in "
+                                          "GeoIP database: {}",
+                                          ip_string, MMDB_strerror(status)));
+        }
+        auto r = field_builder.record();
+        r.field("key", value);
+        r.field("context", output);
+        r.field("timestamp", std::chrono::system_clock::now());
+      } else {
+        field_builder.null();
+      }
+    }
+    return field_builder.finish();
+  }
+
+  /// Inspects the context.
+  auto show() const -> record override {
+    return record{{path_key, db_path_}};
+  }
+
+  /// Updates the context.
+  auto update(table_slice, context::parameter_map)
+    -> caf::expected<record> override {
+    return caf::make_error(ec::unimplemented,
+                           "geoip context can not be updated with events");
+  }
+
+  auto update(chunk_ptr, context::parameter_map)
+    -> caf::expected<record> override {
+    return caf::make_error(ec::unimplemented, "geoip context can not be "
+                                              "updated with bytes");
+  }
+
+  auto update(context::parameter_map parameters)
+    -> caf::expected<record> override {
+    if (parameters.contains(path_key) and parameters.at(path_key)) {
+      db_path_ = *parameters[path_key];
+    }
+    return show();
+  }
+
+  auto save() const -> caf::expected<chunk_ptr> override {
+    // We save the context by formatting into a record of this format:
+    // {path: string}
+    auto builder = flatbuffers::FlatBufferBuilder{};
+    const auto key_key_offset = builder.CreateSharedString(path_key);
+    const auto key_value_offset = pack(builder, data{db_path_});
+    auto field_offsets
+      = std::vector<flatbuffers::Offset<fbs::data::RecordField>>{};
+    field_offsets.reserve(1);
+    const auto record_offset
+      = fbs::data::CreateRecordField(builder, key_key_offset, key_value_offset);
+    field_offsets.emplace_back(record_offset);
+    const auto value_offset
+      = fbs::data::CreateRecordDirect(builder, &field_offsets);
+    const auto data_offset
+      = fbs::CreateData(builder, fbs::data::Data::record, value_offset.Union());
+    fbs::FinishDataBuffer(builder, data_offset);
+    return chunk::make(builder.Release());
```

Comment:
Does this stay just a path to an external file in the medium-term or do we write the mmdb file into the state?

Reply:
The entire `.mmdb` file? The MMDB file sizes I encountered range from 5 to 65 MB, I wonder how large the actual DB sizes would be. I doubt the feasibility of writing the MMDB file contents. The actual DB might be a topic worth exploring.

Reply:
I chained the MMDB lifetime to the context lifetime (with manual updates) in https://github.com/tenzir/tenzir/pull/3731/commits/4f66a26e3324beafc60e58ef4ea99b3005644218. Good enough for now, probably.

Reply:
I loaded the `GeoLite2-City.mmdb` - the file is ~65 MB large, the resulting DB is ~9 MB in memory. It may be feasible to serialize DB content - but not for now.